File: tests/lib/wait.ts
export async function waitForHealth(
  url: string,
  timeoutMs: number = 60_000,
): Promise<void> {
  const start = Date.now();
  let delay = 500;

  while (Date.now() - start < timeoutMs) {
    try {
      const res = await fetch(url);
      if (res.ok) return;
    } catch {
      // not ready yet
    }
    await new Promise((r) => setTimeout(r, delay));
    delay = Math.min(delay * 1.5, 5000);
  }

  throw new Error(`Health check timeout: ${url} not ready after ${timeoutMs}ms`);
}

File: test/e2e/lib/wait.py
from __future__ import annotations

import time

import httpx


def wait_for_healthy(url: str, timeout: float = 60) -> None:
    start = time.monotonic()
    delay = 0.5

    while time.monotonic() - start < timeout:
        try:
            resp = httpx.get(url, timeout=5)
            if resp.is_success:
                return
        except Exception:
            pass
        time.sleep(delay)
        delay = min(delay * 1.5, 5.0)

    raise TimeoutError(f"Health check timeout: {url} not ready after {timeout}s")
